src/tqec/plaquette/rpng_description.py

```diff
@@ -0,0 +1,262 @@
+            raise ValueError('Each plaquette must have 0, 2, or 4 2Q gates.')
+
+
+    @classmethod
+    def from_string(cls, corners_rpng_string: str) -> 'RPNGDescription':
+        """Initialize the RPNGDescription object from a (16+3)-character string"""
+        rpng_objs = tuple([RPNG.from_string(s) for s in corners_rpng_string.split(' ')])
+        if len(rpng_objs) != 4:
+            raise ValueError('There must be 4 corners in the RPNG description.')
+        return cls(rpng_objs)
+    
+
+    @classmethod
+    def from_extended_string(cls, ancilla_and_corners_rpng_string: str) -> 'RPNGDescription':
+        """Initialize the RPNGDescription object from a (16+3)-character string"""
+        values = ancilla_and_corners_rpng_string.split(' ')
+        ancilla_rgn = RGN.from_string(values[0])
+        rpng_objs = tuple([RPNG.from_string(s) for s in values[1:]])
+        if len(rpng_objs) != 4:
+            raise ValueError('There must be 4 corners in the RPNG description.')
+        return cls(rpng_objs, ancilla_rgn)
+
+
+    def get_r_op(self, data_idx: int) -> str | None:
+        """Get the reset operation or Hadamard for the specific data qubit"""
+        return self.corners[data_idx].get_r_op()
+
+    def get_n(self, data_idx: int):
+        """Get the time of the 2Q gate involving the specific data qubit"""
+        return self.corners[data_idx].n
+
+    def get_g_op(self, data_idx: int) -> str | None:
+        """Get the measurement operation or Hadamard for the specific data qubit"""
+        return self.corners[data_idx].get_g_op()
+    
+    def get_plaquette(self, qubits: PlaquetteQubits = SquarePlaquetteQubits()) -> Plaquette:
+        """Get the plaquette corresponding to the RPNG description
+        
+        Note that the ancilla qubit is the last among the PlaquetteQubits and thus
+        has index 4, while the data qubits have indices 0-3.
+        """
+        prep_time = 0
+        meas_time = self.ancilla.n
+        circuit_as_list = [''] * (meas_time - prep_time + 1)
+        for q, rpng in enumerate(self.corners):
+            # 2Q gates.
+            if rpng.n and rpng.p:
+                if rpng.a.value == 'z':
+                    circuit_as_list[rpng.n] += f'C{rpng.p.value.upper()} 4 {q}\n'
+                elif rpng.p.value == 'z':
+                    circuit_as_list[rpng.n] += f'C{rpng.a.value.upper()} {q} 4\n'
+                else:
+                    circuit_as_list[rpng.n] += f'{rpng.a.value.upper()}C{rpng.p.value.upper()} 4 {q}\n'
+            # Data reset or Hadamard.
+            if rpng.r:
+                circuit_as_list[0] += f'{rpng.get_r_op()} {q}\n'
+            # Data measurement or Hadamard.
+            if rpng.g:
+                circuit_as_list[-1] += f'{rpng.get_g_op()} {q}\n'
+        # Ancilla reset and measurement.
+        circuit_as_list[0] += f'R{self.ancilla.r.value.upper()} 4\n'
+        circuit_as_list[-1] += f'M{self.ancilla.g.value.upper()} 4\n'
```

Comment:
See comment on docstring, the hard-coded `4` will introduce holes in qubit indexing for plaquettes on the boundary. I think this is not desirable.

Reply:
Done